engines/optix/cuda/renderer/BasicSimulation.cu

```diff
@@ -0,0 +1,157 @@
+ * This library is free software; you can redistribute it and/or modify it under
+ * the terms of the GNU Lesser General Public License version 3.0 as published
+ * by the Free Software Foundation.
+ *
+ * This library is distributed in the hope that it will be useful, but WITHOUT
+ * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
+ * FOR A PARTICULAR PURPOSE.  See the GNU Lesser General Public License for more
+ * details.
+ *
+ * You should have received a copy of the GNU Lesser General Public License
+ * along with this library; if not, write to the Free Software Foundation, Inc.,
+ * 51 Franklin Street, Fifth Floor, Boston, MA 02110-1301 USA.
+ */
+
+#include <optix_world.h>
+
+struct PerRayData_radiance
+{
+    float3 result;
+    float importance;
+    int depth;
+};
+
+// Scene
+rtDeclareVariable(optix::Ray, ray, rtCurrentRay, );
+rtDeclareVariable(PerRayData_radiance, prd, rtPayload, );
+
+// Material attributes
+rtDeclareVariable(float3, Kd, , );
+
+rtDeclareVariable(float3, geometric_normal, attribute geometric_normal, );
+rtDeclareVariable(float3, shading_normal, attribute shading_normal, );
+
+// Textures
+rtTextureSampler<float4, 2> diffuse_map;
+rtDeclareVariable(float3, texcoord, attribute texcoord, );
+
+// Simulation data
+rtDeclareVariable(unsigned int, use_simulation_data, , );
+rtBuffer<float3> colors;
+rtBuffer<float> opacities;
+rtDeclareVariable(float2, value_range, , );
+rtBuffer<float> simulation_data;
+rtDeclareVariable(unsigned long, user_data, attribute user_data, );
+
+struct InterpolatedValue
+{
+    float influence_low;
+    float influence_high;
+    int idx_low;
+    int idx_high;
+};
+
+// Finds between what pair of values a given value is and how much influence
+// each of these values have
+static __device__ inline InterpolatedValue interpolateValues(
+    const float v_min, const float v_max, const float value,
+    const int num_values)
+{
+    InterpolatedValue result;
+
+    const float v_clamped = min(v_max, max(v_min, value));
+    const float range_per_value = (v_max - v_min) / (num_values - 1);
+    const float idx_value = (v_clamped - v_min) / range_per_value;
+
+    // Indices of the two values to interpolate between
+    const int idx_value_low = max(0, int(floor(idx_value)));
+    const int idx_value_high = min(num_values - 1, int(ceil(idx_value)));
+
+    result.idx_low = idx_value_low;
+    result.idx_high = idx_value_high;
+
+    if (idx_value_low == idx_value_high)
+    {
+        // Since both values are the same we split the influence
+        result.influence_low = 0.5f;
+        result.influence_high = 0.5f;
+    }
+    else
+    {
+        const float v_low = v_min + float(idx_value_low) * range_per_value;
+        const float v_high = v_min + float(idx_value_high) * range_per_value;
+
+        result.influence_low = 1.0f - abs(v_clamped - v_low) / range_per_value;
+        result.influence_high =
+            1.0f - abs(v_clamped - v_high) / range_per_value;
```

Comment:
You don't need two values, if you're **linearly** interpolating between two points, influence_low = 1 - influence_high.